### PR TITLE
Set up testing and wheels for Windows on GitHub Actions

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

Lays the groundwork for testing and wheel building for Windows on GitHub Actions.

## Related issue number

#631

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: ci/travis/install-dependencies.sh [hunk 1]
```diff
@@ -141,13 +172,20 @@ install_dependencies() {
...
     install_miniconda
-    pip install flake8==3.7.7 flake8-comprehensions flake8-quotes==2.0.0  # Python linters
-    pushd "${WORKSPACE_DIR}/python/ray/dashboard/client"  # TypeScript & HTML linters
-      install_node
-      run_npm
-    popd
+    pip install flake8==3.7.7 flake8-comprehensions flake8-quotes==2.0.0 yapf==0.23.0  # Python linters
+    # readthedocs has an antiquated build env.
+    # This is a best effort to reproduce it locally to avoid doc build failures and hidden errors.
+    local filename
```
Comment (pcmoritz) [inline ci/travis/install-dependencies.sh:182]: why is the below change needed? pip install seems to have worked before just fine

File: ci/travis/install-dependencies.sh [hunk 2]
```diff
@@ -121,6 +123,35 @@ run_npm() {
...
+  if [ "${OSTYPE}" = msys ]; then
+    # Not Windows-compatible: https://github.com/npm/cli/issues/558#issuecomment-584673763
+    { echo "WARNING: Skipping running NPM due to module incompatibilities with Windows"; } 2> /dev/null
+  else
+    run_npm
+  fi
+}
+
+# Takes a list of pip requirements and prints the ones that don't have known incompatibilities with the current system.
+filter_compatible_pip_requirements() {
```
Comment (pcmoritz) [inline ci/travis/install-dependencies.sh:136]: Let's remove this. We can for example address it later by trying to remove the dependencies that are incompatible.

File: ci/travis/ci.sh [hunk 3]
```diff
@@ -38,21 +38,187 @@ should_run_job() {
...
+lint_bazel() {
+  local platform="" architecture="" suffix=".tar.gz"
+  case "${OSTYPE}" in
+    linux*) platform=linux;;
+    darwin*) platform=darwin;;
+    msys*) platform=windows; suffix=".zip";;
+  esac
+  case "${HOSTTYPE}" in
+    x86_64) architecture=amd64;;
+    i?86*) architecture=i386;;
```
Comment (pcmoritz) [inline ci/travis/ci.sh:146]: Why do we need to support that?

File: ci/travis/ci.sh [hunk 4]
```diff
@@ -38,21 +38,187 @@ should_run_job() {
...
+  # PATH for all platforms
+  paths+=(
+    "${HOME}/miniconda/bin"
+  )
+  # PATH for Windows
+  test "${OSTYPE}" != msys || paths+=(
+    "${HOME}/miniconda/bin/Scripts"
+  )
+
+  export "${definitions[@]}" 2> /dev/null
```
Comment (pcmoritz) [inline ci/travis/ci.sh:77]: why do this delayed? Why not just exporting stuff directly above?

File: ci/travis/ci.sh [hunk 5]
```diff
@@ -38,21 +38,187 @@ should_run_job() {
...
+
+  prepend_PATH "${paths[@]}"
```
Comment (pcmoritz) [inline ci/travis/ci.sh:79]: same

File: ci/travis/ci.sh [hunk 6]
```diff
@@ -38,21 +38,187 @@ should_run_job() {
...
+}
+
+# Prepends each argument to PATH if it isn't already in the PATH.
+prepend_PATH() {
```
Comment (pcmoritz) [inline ci/travis/ci.sh:83]: try to get rid of this

File: ci/travis/ci.sh [hunk 7]
```diff
@@ -38,21 +38,187 @@ should_run_job() {
...
   fi
   test -z "${set_x}" || set -x  # restore set -x
+}
+
+test_python() {
   if [ "${OSTYPE}" = msys ]; then
-    PATH="${PATH// :/:}"  # HACK: Work around https://github.com/actions/virtual-environments/issues/635#issue-589318442
+    # Windows -- most tests won't work yet; just do the ones we know work
+    PYTHONPATH=python python -m pytest --durations=5 --timeout=300 python/ray/tests/test_mini.py
+  fi
```
Comment (pcmoritz) [inline ci/travis/ci.sh:109]: what about non windows?

File: ci/travis/ci.sh [hunk 8]
```diff
@@ -38,21 +38,187 @@ should_run_job() {
...
+}
+
+lint_python() {
+  # ignore dict vs {} (C408), others are defaults
+  command -V python
+  python -m flake8 --inline-quotes '"' --no-avoid-escape --exclude=python/ray/core/generated/,streaming/python/generated,doc/source/conf.py,python/ray/cloudpickle/,python/ray/thirdparty_files --ignore=C408,E121,E123,E126,E226,E24,E704,W503,W504,W605
+  "${ROOT_DIR}"/format.sh --all
+}
+
+lint_bazel() {
```
Comment (pcmoritz) [inline ci/travis/ci.sh:137]: it is enough to lint bazel on the linux build!

File: ci/travis/ci.sh [hunk 9]
```diff
@@ -38,21 +38,187 @@ should_run_job() {
...
+
+    # Build buildifier
+    go get github.com/bazelbuild/buildtools/buildifier
+
+    # Now run buildifier
+    "${ROOT_DIR}"/bazel-format.sh
+  )
+}
+
+lint_web() {
```
Comment (pcmoritz) [inline ci/travis/ci.sh:173]: let's keep the linting code we had before here (especially, get rid of all the if things here)

File: ci/travis/ci.sh [hunk 10]
```diff
@@ -127,24 +313,90 @@ build() {
     eval "$(curl -sL https://raw.githubusercontent.com/travis-ci/gimme/master/gimme | GIMME_GO_VERSION=stable bash)"
   fi
 
-  if [ "${LINUX_WHEELS-}" = 1 ]; then
-    # Mount bazel cache dir to the docker container.
-    # For the linux wheel build, we use a shared cache between all
-    # wheels, but not between different travis runs, because that
-    # caused timeouts in the past. See the "cache: false" line below.
-    local MOUNT_BAZEL_CACHE=(-v "${HOME}/ray-bazel-cache":/root/ray-bazel-cache -e TRAVIS=true -e TRAVIS_PULL_REQUEST="${TRAVIS_PULL_REQUEST}" -e encrypted_1c30b31fe1ee_key="${encrypted_1c30b31fe1ee_key-}" -e encrypted_1c30b31fe1ee_iv="${encrypted_1c30b31fe1ee_iv-}")
+  case "${OSTYPE}" in
```
Comment (pcmoritz) [inline ci/travis/ci.sh:316]: No need to check OSTYPE here, we just need to make sure that `LINUX_WHEELS`, `MAC_WHEELS` and `WINDOWS_WHEELS` are only defined on the right OS.

File: ci/travis/ci.sh [hunk 11]
```diff
@@ -127,24 +313,90 @@ build() {
...
-    suppress_output docker run --rm -w /ray -v "${PWD}":/ray "${MOUNT_BAZEL_CACHE[@]}" -e TRAVIS_COMMIT="${TRAVIS_COMMIT}" -ti rayproject/arrow_linux_x86_64_base:python-3.8.0 /ray/python/build-wheel-manylinux1.sh
+  if [ "${TESTSUITE-}" = cpp_worker ]; then
+    bazel test --config=ci //cpp:all --build_tests_only --test_output=streamed
   fi
 
-  if [ "${MAC_WHEELS-}" = 1 ]; then
-    # This command should be kept in sync with ray/python/README-building-wheels.md.
-    suppress_output "${WORKSPACE_DIR}"/python/build-wheel-macos.sh
+  local should_test_wheels=""
+  case "${OSTYPE}" in
```
Comment (pcmoritz) [inline ci/travis/ci.sh:368]: same here

File: ci/travis/ci.sh [hunk 12]
```diff
@@ -127,24 +313,90 @@ build() {
...
+    linux*) should_test_wheels="${LINUX_WHEELS-}";;
+    darwin*) should_test_wheels="${MAC_WHEELS-}";;
+  esac
+  if [ "${should_test_wheels}" = 1 ]; then
+    "${WORKSPACE_DIR}"/ci/travis/test-wheels.sh || { result=$? && flush_logs=1; }
+  fi
+
+  ##### END TASKS #####
```
Comment (pcmoritz) [inline ci/travis/ci.sh:376]: what is this? We should get rid of it.

File: ci/travis/ci.sh [hunk 13]
```diff
@@ -127,24 +313,90 @@ build() {
...
+  return "${result}"
+}
+
+_main() {
+  if [ -n "${GITHUB_WORKFLOW-}" ]; then
+    # Necessary for GitHub Actions (which uses separate shells for different commands)
+    # Unnecessary for Travis (which uses one shell for different commands)
+    reload_env
+  fi
+  if [ 0 -lt "$#" ]; then
```
Comment (pcmoritz) [inline ci/travis/ci.sh:395]: what does this do?

File: ci/travis/ci.sh (hunk 10 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:316]: In terms of the design I actually think it should be the other way around; the redundant bit is those flags. (They should just be just a single `WHEELS` flag.)

I can look into doing it the other way around, though I'm not sure how feasible it is, since it depends somewhat on what kinds of conditoinal expressions we can put as the definitions of these variables.

File: ci/travis/ci.sh (hunk 12 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:376]: You mean the comment? It's just indicating that all the build tasks need to go before it. Because the step after needs to come last (flushing the CI logs), otherwise someone might add stuff after that by accident?

File: ci/travis/ci.sh (hunk 13 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:395]: It's supposed to check if any arguments were provided before attempting to execute the provided command, but it might be necessary now, I can probably remove it.

File: ci/travis/ci.sh (hunk 8 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:137]: True, assuming we have a Linux build running on GHA. If for some reason we decide to turn that off later (say, if we see it consumes too many minutes) then this would be quite beneficial to run on Windows since it'll help me and everyone else see lint errors before a PR is opened. Do you still want me to take it out?

File: ci/travis/ci.sh (hunk 7 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:109]: I haven't migrated those tests quite yet, they're still running on Travis right now.

File: ci/travis/ci.sh (hunk 6 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:83]: I'll look into it, but it may have repercussions and require a *lot* of refactoring to ensure we're spawning a new shell after environment modifications...

File: ci/travis/ci.sh (hunk 4 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:77]: No strict reason, I just did it out of habit since it separates out the export command (imperative) from the declarative portion. I can change it if you prefer to do it directly above.

File: ci/travis/ci.sh (hunk 3 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:146]: I guess we don't; I just added the `case` to make it clear there's a translation from `HOSTTYPE` and then this seemed like a cheap line to add afterwards. I can remove it.

File: ci/travis/install-dependencies.sh (hunk 2 above)
Comment (mehrdadn) [reply ci/travis/install-dependencies.sh:136]: Wouldn't it just break something if I remove them from the requirements file? I think one of them was [`blist`](https://github.com/DanielStutzbach/blist/issues/81#issue-391460716), I forget if there was another one. Should I just delete those lines from the requirements? I would've guessed it would break something for users.

File: ci/travis/install-dependencies.sh (hunk 1 above)
Comment (mehrdadn) [reply ci/travis/install-dependencies.sh:182]: The `sed` line I can probably remove (I think I added it a while ago to help me debug some stuff), but the `filter_compatible_pip_requirements` line is needed since the `blist` requirement doesn't work on on Windows as I mentioned below.

File: ci/travis/ci.sh (hunk 10 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:316]: Yeah so like I expected it doesn't seem like there's a reasonable way to do this in the YAML file since e.g. `LINUX_WHEELS: ${{ runner.os == 'Linux' }` turns into `LINUX_WHEELS=true` rather than `LINUX_WHEELS=1`. Not sure how you want me to do it otherwise, it seems any other way would be less elegant?

File: .travis.yml [hunk 14]
```diff
@@ -308,21 +278,20 @@ script:
...
-  - cd ../..
+  - (cd deploy/ray-operator && export CC=gcc && suppress_output go build && suppress_output go test ./...)
 
   # random python tests TODO(ekl): these should be moved to bazel
   - if [ $RAY_CI_PYTHON_AFFECTED == "1" ]; then python -m pytest -v --durations=5 --timeout=300 python/ray/experimental/test/async_test.py; fi
 
   # bazel python tests. This should be run last to keep its logs at the end of travis logs.
-  - if [ $RAY_CI_PYTHON_AFFECTED == "1" ]; then ./ci/keep_alive bazel test --config=ci --test_tag_filters=-jenkins_only python/ray/tests/...; fi
-  - if [ $RAY_CI_TUNE_AFFECTED == "1" ]; then ./ci/keep_alive bazel test --config=ci --test_tag_filters=-jenkins_only python/ray/tune/...; fi
+  - if [ $RAY_CI_PYTHON_AFFECTED == "1" ]; then bazel_targets+=(python/ray/tests/...); fi
```
Comment (pcmoritz) [inline .travis.yml:292]: Why is this change needed?

File: ci/travis/ci.sh (hunk 8 above)
Comment (pcmoritz) [reply ci/travis/ci.sh:102]: Yes!

File: .travis.yml (hunk 14 above)
Comment (mehrdadn) [reply .travis.yml:292]: It factors out and deduplicates the `bazel test` calls for maintainability.

File: ci/travis/ci.sh (hunk 10 above)
Comment (pcmoritz) [reply ci/travis/ci.sh:264]: Why not just set ``WHEELS_TYPE: ${{ runner.os }}`` and then check for that here?

File: ci/travis/ci.sh [hunk 15]
```diff
@@ -38,21 +38,142 @@ should_run_job() {
...
+  fi
+
+  # NOTE: Modifying PATH invalidates Bazel's cache! Do not add to PATH unnecessarily.
+  PATH="${HOME}/miniconda/bin":"${PATH}"
+  if [ "${OSTYPE}" = msys ]; then
+    PATH="${HOME}/miniconda/bin/Scripts":"${PATH}"
   fi
-  test -z "${set_x}" || set -x  # restore set -x
+
+  # Deduplicate PATH
```
Comment (pcmoritz) [inline ci/travis/ci.sh:66]: This should not be needed...

File: ci/travis/ci.sh (hunk 10 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:264]: Well I was also trying not to modify how the existing code works at the same time as I'm setting up GHA, but that solution also seems redundant and different from how we handle other flags? (And it suggests the wheel type can be different from the current OS which would be misleading.)

File: ci/travis/ci.sh (hunk 12 above)
Comment (pcmoritz) [reply ci/travis/ci.sh:325]: No, I mean the log flushing from below!

File: ci/travis/ci.sh (hunk 15 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:66]: What are you basing this on?

`reload_env` can (and I believe is) called multiple times in a single shell. If you change `PATH` *n* times you're going to build with Bazel *n* times. I don't think that's a good idea.

File: ci/travis/ci.sh (hunk 12 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:325]: I didn't add that, it used to be there. I just moved it to this file. But I'm pretty sure it's necessary.

File: ci/travis/install-dependencies.sh (hunk 1 above)
Comment (pcmoritz) [reply ci/travis/install-dependencies.sh:180]: I think we can remove it from the `requirements-rtd`, because we are mocking it for read the docs anyways. Does that work?

Comment (mehrdadn) [reply ci/travis/install-dependencies.sh:180]: Not sure, I can give it a shot and see.

File: .travis.yml (hunk 14 above)
Comment (pcmoritz) [reply .travis.yml:292]: No, it makes it harder to read (and harder to copy and paste into a terminal if you want to run it.

File: ci/travis/install-dependencies.sh (hunk 1 above)
Comment (pcmoritz) [reply ci/travis/install-dependencies.sh:180]: Yeah, I'm pretty sure it will work :)

File: .travis.yml (hunk 14 above)
Comment (mehrdadn) [reply .travis.yml:292]: Hmm okie

File: .travis.yml [hunk 16]
```diff
@@ -150,8 +122,7 @@ matrix:
         - . ./ci/travis/ci.sh build
       script:
         - . ./ci/travis/ci.sh preload
-        # Explicitly sleep 60 seconds for logs to go through
-        - ./ci/travis/test-wheels.sh || { cat /tmp/ray/session_latest/logs/* && sleep 60 && false; }
+        - . ./ci/travis/ci.sh run
```
Comment (pcmoritz) [inline .travis.yml:125]: This should be `. ./ci/travis/ci.sh test_wheels`

File: .travis.yml [hunk 17]
```diff
@@ -167,8 +138,7 @@ matrix:
...
```
Comment (pcmoritz) [inline .travis.yml:141]: This should be `. ./ci/travis/ci.sh test_wheels`

File: .travis.yml [hunk 18]
```diff
@@ -297,7 +267,7 @@ matrix:
         - . ./ci/travis/ci.sh build
       script:
         - . ./ci/travis/ci.sh preload
-        - bazel test --config=ci //cpp:all --build_tests_only --test_output=streamed
+        - . ./ci/travis/ci.sh run
```
Comment (pcmoritz) [inline .travis.yml:270]: This should be `. ./ci/travis/ci.sh test_cpp`

**Review by pcmoritz — APPROVED** [review]
